community/kernel/src/main/java/org/neo4j/graphdb/notification/Notification.java

```diff
@@ -0,0 +1,153 @@
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.graphdb.notification;
+
+/**
+ * Representation for notifications found when executing a query.
+ *
+ * A notification can be visualized in a client pinpointing problems or other information about the query.
+ */
+public final class Notification
+{
+    private final NotificationCode code;
+    private final String title;
+    private final String description;
+    private final NotificationSeverityLevel severity;
+    private final InputPosition position;
+
+    /**
+     * use {@link Builder}
+     */
+    private Notification( NotificationCode code, String title, String description, NotificationSeverityLevel severity, InputPosition position )
+    {
+        this.code = code;
+        this.title = title;
+        this.description = description;
+        this.severity = severity;
+        this.position = position;
+    }
+
+    /**
+     * Returns a notification code for the discovered issue.
+     * @return the notification code
+     */
+    public NotificationCode getCode()
+    {
+        return code;
+    }
+
+    /**
+     * Returns a short summary of the notification.
+     * @return the title of the notification.
+     */
+    public String getTitle()
+    {
+        return title;
+    }
+
+    /**
+     * Returns a longer description of the notification.
+     * @return the description of the notification.
+     */
+    public String getDescription()
+    {
+        return description;
+    }
+
+    /**
+     * Returns the severity level of this notification.
+     * @return the severity level of the notification.
+     */
+    public NotificationSeverityLevel getSeverity()
+    {
+        return severity;
+    }
+
+    /**
+     * The position in the query where this notification points to.
+     * Not all notifications have a unique position to point to and should in
+     * that case return {@link InputPosition#empty}
+     *
+     * @return the position in the query where the issue was found, or
+     * {@link InputPosition#empty} if no position is associated with this notification.
+     */
+    public InputPosition getPosition()
+    {
+        return position;
+    }
+
+    /**
+     * Builds a notification message.
+     */
+    public static class Builder
```

Comment:
If the severity and position are the only aspects that are customisable, and if the `code`, `title`, and `description` are all part of the `NotificationCode` ([see other comment](https://github.com/neo4j/neo4j/pull/4135/files#r26231767)). Then perhaps the builder could be greatly simplified, possibly even replaced with factory methods on the `NotificationCode`.

Either like this:

    enum NotificationCode {
        // actual codes go here...
        ;
        private final String code, title, description;
        private NotificationCode( String code, String title, String description ) { /*do the obvious*/ }
        public Notification notification( InputPosition position, SeverityLevel level ) {
            return new Notification( this, position, level );
        }
        public Notification notification( InputPosition position ) { return notification( position, SeverityLevel.INFO ); }
        public Notification notification( SeverityLevel level ) { return notification( InputPosition.empty, level ); }
        public Notification notification() { return notification( InputPosition.empty, SeverityLevel.INFO ); }
        public String getCode() { return code; }
        public String getTitle() { return title; }
        public String getDescription() { return description; }
    }

or like this (with the `Notification.Builder` constructor being package-private):

    enum NotificationCode {
        // actual codes go here...
        ;
        private final String code, title, description;
        private NotificationCode( String code, String title, String description ) { /*do the obvious*/ }
        public Notification notification( InputPosition position, SeverityLevel level ) {
            return new Notification( this, position, level );
        }
        public Notification.Builder notification() { return new Notification.Builder( this ); }
    }